codegen/generator/go/generator.go

```diff
@@ -45,8 +66,28 @@ func (g *GoGenerator) Generate(_ context.Context, schema *introspection.Schema)
 			return nil
 		},
 		Object: func(t *introspection.Type) error {
+			objectTmpl := templates.Object
+
+			// don't create methods on query for the env itself, only its deps
+			// e.g. don't create `func (r *DAG) Go() *Go` in the Go env's codegen
+			if g.Config.EnvironmentName != "" && t.Name == generator.QueryStructName {
```

Comment:
Interesting, so without this check would user envs have all of their entrypoints reflected on their own `dag` var after codegen?

Makes sense, but devil's advocate: theoretically it could be useful to call yourself through the DAG to leverage the entrypoint-level caching, instead of directly calling a sibling function. Would that "just work?" I'm not sure if it's worth exposing anyway, since it's a bit :exploding_head:. So far I've been able to write mostly pure entrypoints that just chain API calls and don't return `error`, so in that case the entrypoint-level caching would be pure overhead.

BTW: is there anything stopping someone from creating an env called `Container` and overriding the internal API?

Reply:
> Interesting, so without this check would user envs have all of their entrypoints reflected on their own dag var after codegen?

Sort of, we actually stop that from happening in most cases much earlier than this code by only loading the env deps into the schema when generating for a given env.

I think this branch of code is leftover from the Zenith branch where we had universe, which was special because it was intrinsically loaded along side `core`. I'll try deleting this code from this branch to confirm.

---

Either way, yes totally agree being able to self call would be nice. The problem is the catch-22s you get yourself into when you're in the middle of writing env code, need to re-run codegen, but then it fails because your code doesn't compile. We ran into that in cloak and it was pretty maddening.

I would like to have this ability someday, but I think we'll need a radically different approach. My idea is that since we're already doing codegen + ast parsing, we can literally just take the user code and wrap it up such that when it's actually compiled, any direct call to another env entrypoint becomes wrapped in a call that goes through the actual server api (and thus wrapped in caching).

I'm pretty sure that isn't *quite* as crazy as it sounds; I think there's approaches to it that might actually be fairly robust. But it's still *sort of* crazy and thus something I'm not pursuing for now 🙂

---

> BTW: is there anything stopping someone from creating an env called Container and overriding the internal API?

Yes, that should result in an error because there will be a conflict during the schema merging. Sounds like a good test case though to be sure :-)

Reply:
> theoretically it could be useful to call yourself through the DAG to leverage the entrypoint-level caching, instead of directly calling a sibling function.

If I understand you correctly I plan on supporting this in Python, without needing or leveraging codegen:

https://github.com/dagger/dagger/blob/431ae2f5614596cb0a1ea779abee33319ffd07f7/sdk/python/src/dagger/server/_resolver.py#L271

General idea is this:

```python
from dagger.ext import function

@function
async def build():
    ...

@function
async def test():
    await build()  # calls function directly
    # or
    await build.call()  # calls function via API/DAG
    ...
```

This is possible because the decorator can return an object that wraps the function and it already has everything it needs to use the query builder underneath without relying on codegen.

There's precedent for this pattern. For example, in [Celery](https://docs.celeryq.dev/en/stable/userguide/calling.html#basics) you can do `my_task()` to call a function in the current process vs `my_task.apply_async()` to send it to a worker.

In [Prefect](https://docs.prefect.io/2.11.5/concepts/tasks/#tasks-overview) it's the opposite convention. You do `my_task.fn()` to call the function directly vs `my_task()` to run it in the executor. Also `my_task.submit()` for running async.

Reply:
Makes total sense @helderco, this is definitely another area where I think we should be okay with each SDK having it's own sort of approach.

In Go, obviously nothing like that is really possible without adding overhead to the actual code the user writes (which I want to avoid at all costs, e.g. having them wrap `funcs` in some custom type that extends it with a method to call directly or with caching), so I think our only option there is to be more magical with some sort of codegen approach that rewrites funcs with that wrapping inherently.

Reply:
You're talking about always going through the API right? I was thinking on defaulting to a direct call since it seems less surprising, and explicitly requiring a `my_func.call()` to go through the API. That's the same as doing `dagger.call(my_func)`, which can be done in Go as well.

You think defaulting to go through the API makes the most sense (when making a direct call)? I would want to think so, but seems like more surprising to users. It's just a matter of documentation though (like with Prefect).

However it does make a weird case if I create a `@function` that uses only lazy API methods (not async), because going through the API forces it to be async so there's a signature mismatch that's easier to enforce with the added `my_task.call()` method, or external `call(my_task)`.

Should be the same in Go, requiring ctx and error.